### PR TITLE
Handle 206

### PR DESCRIPTION
**Pull request by heyams**

### CODE REVIEW

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineResponse.java [hunk 1]
```diff
@@ -33,28 +33,41 @@ public String getBody() {
...
         return errorNodes.stream()
-            .map(errorNode -> errorNode.get("message").asText())
-            .filter(s -> !s.equals("Telemetry sampled out."))
+            .map(errorNode -> new ResponseError(errorNode.get("index").asInt(), errorNode.get("statusCode").asInt(), errorNode.get("message").asText()))
+            .filter(s -> !s.getMessage().equals("Telemetry sampled out."))
             .collect(Collectors.toSet());
     }
+
+    @SuppressWarnings("SystemOut")
+    public static void main(String[] args) {
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineResponse.java:66]: test will be moved to java agent smoke test

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java [hunk 2]
```diff
@@ -176,41 +150,6 @@ private TelemetryItem createOtelResourceMetric(TelemetryItemBatchKey telemetryIt
...
-    }
-
-    private static void writeTelemetryItems(JsonGenerator jg, List<TelemetryItem> telemetryItems)
-        throws IOException {
-        jg.setRootValueSeparator(new SerializedString("\n"));
-        for (TelemetryItem telemetryItem : telemetryItems) {
-            mapper.writeValue(jg, telemetryItem);
-        }
-    }
-
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: why extract these to a new class?

Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: reusability

Comment (trask) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: maybe I missed, but which methods are being used by more than one (non-test) class?

Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: all the public methods are used more than once. feel free to check out the PR and verify yourself.

Comment (trask) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: ok, this is what I found:

* PipelineUtil.deserializeTelemetryItem is only used by LocalStorageTelemetryPipelineListener
* PipelineUtil.decode is only used by LocalStorageTelemetryPipelineListener

Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: and tests

and 
image

Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: it's my preference.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/AppInsightsByteBufferPool.java [hunk 3]
```diff
@@ -8,7 +8,7 @@
 import java.util.Queue;
 import java.util.concurrent.ArrayBlockingQueue;
 
-class AppInsightsByteBufferPool {
+public class AppInsightsByteBufferPool {
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/AppInsightsByteBufferPool.java:11]: is this change needed?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/ByteBufferOutputStream.java [hunk 4]
```diff
@@ -8,15 +8,15 @@
...
+public class ByteBufferOutputStream extends OutputStream {
 
     private final AppInsightsByteBufferPool byteBufferPool;
 
     private final List<ByteBuffer> byteBuffers = new ArrayList<>();
 
     private ByteBuffer current;
 
-    ByteBufferOutputStream(AppInsightsByteBufferPool byteBufferPool) {
+    public ByteBufferOutputStream(AppInsightsByteBufferPool byteBufferPool) {
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/ByteBufferOutputStream.java:19]: are these changes needed?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java (hunk 2 above)
Comment (trask) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemExporter.java:213]: for methods that are only called from a single (non-Test) class, I'd recommend collocating them with the code that calls them, as it makes it easier for other people to read and follow

if the class was very big, that could be a good reason to violate this rule and split the code out, but that doesn't seem to be the case here

if you need to make them visible for tests, you can then mark them as [`// visible for testing`](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-java-instrumentation%20visible%20for%20testing&type=code)

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/ByteBufferOutputStream.java (hunk 4 above)
Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/ByteBufferOutputStream.java]: reverted.. earlier change requires this to be public.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/AppInsightsByteBufferPool.java (hunk 3 above)
Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/AppInsightsByteBufferPool.java]: same here

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemSerialization.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    public static List<TelemetryItem> deserialize(byte[] data) {
+        return deserializeWithoutEncode(decode(data));
+    }
+
+    // visible for testing
+    // deserialize raw bytes to a list of TelemetryItem without decoding
+    public static List<TelemetryItem> deserializeWithoutEncode(byte[] data) {
+        try {
+            ObjectMapper mapper = createObjectMapper();
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemSerialization.java:39]: this line isn't needed (can use existing static field `mapper`)
```suggestion
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineResponse.java [hunk 6]
```diff
@@ -33,28 +33,32 @@ public String getBody() {
...
 
-    private static Set<String> parseErrors(String body) {
+    static Set<ResponseError> parseErrors(String body) {
         JsonNode jsonNode;
         try {
             jsonNode = new ObjectMapper().readTree(body);
         } catch (JsonProcessingException e) {
             // fallback to generic message
-            return singleton("Could not parse response");
+            return singleton(null);
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineResponse.java:55]: I think we're losing some logging that we used to get from this

maybe let `JsonProcessingException` bubble up from this method and handle it in `getErrorMessage` to return `singleton("Could not parse response")`

and in `getErrors` can ignore it and return `emptSet()`

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/localstorage/LocalStorageTelemetryPipelineListener.java [hunk 7]
```diff
@@ -46,16 +58,40 @@ public LocalStorageTelemetryPipelineListener(
...
-                request.getConnectionString(), request.getTelemetry(), getOriginalErrorMessage(response));
+                request.getConnectionString(), request.getByteBuffers(), getOriginalErrorMessage(response));
+        } else if (statusCode == 206) {
+            processStatusCode206(request, response);
+        }
+    }
+
+    private void processStatusCode206(TelemetryPipelineRequest request, TelemetryPipelineResponse response) {
+        Set<ResponseError> errors = response.getErrors();
+        errors.forEach(error -> logger.verbose("Error in telemetry: " + error));
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/localstorage/LocalStorageTelemetryPipelineListener.java:72]: (to avoid the unnecessary allocations when verbose logging is not enabled)
```suggestion
        errors.forEach(error -> logger.verbose("Error in telemetry: {}", error));
```

and/or surround in if verbose enabled conditional

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/ResponseError.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    public String getMessage() {
+        return message;
+    }
+
+    public String toString() {
+        return "ErrorResponse{" +
+            "index=" + index +
+            ", statusCode=" + statusCode +
+            ", message='" + message + '\'' +
+            '}';
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/ResponseError.java:34]:
```suggestion
            ", message='" + message + "'}";
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemSerialization.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    private static final AppInsightsByteBufferPool byteBufferPool = new AppInsightsByteBufferPool();
+    private static final ClientLogger logger = new ClientLogger(TelemetryItemSerialization.class);
+
+    public static List<TelemetryItem> deserialize(byte[] data) {
+        return deserializeWithoutEncode(decode(data));
+    }
+
+    // visible for testing
+    // deserialize raw bytes to a list of TelemetryItem without decoding
+    public static List<TelemetryItem> deserializeWithoutEncode(byte[] data) {
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemSerialization.java:37]:
```suggestion
    public static List<TelemetryItem> deserializeAlreadyDecoded(byte[] data) {
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemSerialization.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    private static void writeTelemetryItems(JsonGenerator jg, List<TelemetryItem> telemetryItems, ObjectMapper mapper)
+        throws IOException {
+        jg.setRootValueSeparator(new SerializedString("\n"));
+        for (TelemetryItem telemetryItem : telemetryItems) {
+            mapper.writeValue(jg, telemetryItem);
+        }
+    }
+    private TelemetryItemSerialization() {
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryItemSerialization.java:110]:
```suggestion

    private TelemetryItemSerialization() {
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineResponse.java [hunk 11]
```diff
@@ -33,28 +34,39 @@ public String getBody() {
...
+        } catch (IllegalStateException e) {
+            return emptySet();
+        }
+    }
+
+    public Set<String> getErrorMessages() {
+        Set<ResponseError> responseErrors = getErrors();
+        if (responseErrors.isEmpty()) {
+            return singleton("Could not parse response");
+        }
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/pipeline/TelemetryPipelineResponse.java:49]: (in case there really are no errors)
```suggestion
        Set<ResponseError> responseErrors;
        try {
            return parseErrors(body);
        } catch (IllegalStateException e) {
            return singleton("Could not parse response");
        }
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/utils/TestUtils.java [hunk 12]
```diff
@@ -27,8 +27,12 @@ public final class TestUtils {
             + "IngestionEndpoint=https://test.in.applicationinsights.azure.com/;"
             + "LiveEndpoint=https://test.livediagnostics.monitor.azure.com/";
 
+    public static TelemetryItem createMetricTelemetry(String name, int value, String connectionString) {
+        return createMetricTelemetry(name, value, connectionString, null, null);
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/utils/TestUtils.java:31]: then you don't need the if/else below
```suggestion
        return createMetricTelemetry(name, value, connectionString, "state", "blocked");
```

**Review by trask — APPROVED** [review]
👍
